### ISSUE TITLE
ui Elements, new attr root_view - enhancement 

### ISSUE THREAD

**Issue by Phuket2** [omz/Pythonista-Issues#539]
It occurs to me that if all ui elements had an attr named something like 'root_view' it would be worth the extra bytes to store the most top enclosing view.  I realise you can traverse the v.subviews to go up and down the view tree. But for eg. I have a btn buried 3 or 4 subviews deep ands lets say that btn is to close the view. Then your action func has no context really i.e superview has a no meaning generically.  
Ok, maybe a method in the ui module to get the root view is a better solution. Either way, it seems like if this functionality was built into the ui module, you could code more explicitly when you have a lot of nested subviews in a view. 

To extend the idea a little bit, maybe the sort of functionality that the below class provides as utility functions could be really helpful to people using the ui.

***The below is based on code @JonB wrote, in response to some forum questions.  The code that works will be his, if something is not working that will be my edits.***

```python 
import ui
from collections import deque, namedtuple


class ViewWalker(object):
    # simple iterator for ui.View objects,
    # capable of depth or breadth first traversal
    def __init__(self, v, breadthfirst=False):
        self._dq = deque([v])
        self._breadth = breadthfirst

    def __iter__(self):
        # required for iterator objects'''
        return self

    def root_view(self):
        # silly attempt to get the root view
        return self.as_list()[0]

    def __next__(self):
        # required for iterator objects.
        # raise stopiteration once the queue is empty.
        if not self._dq:
            raise StopIteration
        # pop next view...
        if self._breadth:
            v = self._dq.popleft()  # oldest entry (FIFO)
        else:
            v = self._dq.pop()  # newest entry (stack)
        # then push its subviews
        if hasattr(v, 'subviews'):
            self._dq.extend(v.subviews)
        return v

    def as_list(self, breadthfirst=False):
        self._breadth = breadthfirst
        return [s for s in self]

    def as_list_names(self, breadthfirst=False):
        self._breadth = breadthfirst
        return [s.name for s in self]

    def as_dict(self, breadthfirst=False):
        # returns a dict
        self._breadth = breadthfirst
        return {s.name: s for s in self}

    def as_namedtuple(self, breadthfirst=False):
        # returns a namedtuple
        self._breadth = breadthfirst
        d = self.as_dict(breadthfirst)
        return namedtuple('ViewObjects', d.keys())(**d)
```

**Comment by jsbain** [comment]
While I agree this is something I'd use, going up is pretty easy (except for ButtonItems)
``` 
def find_root(v):
   root=v
   while root.superview:
       root=root.superview
   return root

 ...
def action(sender):
      root=find_root(sender)
...
```

**Comment by Phuket2** [comment]
@jsbain , fair enough. I get your point also.  Rightly or wrongly it seems to me adding some of these ultility functions to the ui module would provide more clarity to how things work. Eg. You are reading through the documentation and you see a find_root method/func, then you read about it in the docs and then some light bulbs go off about nested views.  Or lets say it reinforces some of the other documentation that maybe your eyes skipped over.  
Regarding ui.ButtonItems and ui.NavigationView It seems to me that it would be great if they were like all other ui Elements with the atttrs set accordingly so you could just iterate over the objects without having to special case them.  But I except this could be a naive point of view(over simplification).  Just seems more cohesive

**Comment by Phuket2** [comment]
@jsbain, i also just realise it was so stupid of me to try and extend your ViewWalker code to be a general helper class.  Does not make sense as the iterator can not be reset. I can see there are some ways with itertools.tee, which dont make sense for this use case.  Feel stupid I didn't get this after so long. Live and learn, i guess!
